pkg/metrics/collector/snapshotter.go

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2022. Nydus Developers. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package collector
+
+import (
+	"fmt"
+	"os/exec"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/nydus-snapshotter/pkg/metrics/data"
+)
+
+type SnapshotterMetricsCollector struct{}
+type SnapshotterCacheUsageCollector struct{}
+
+var GlobalSnapshotterMetricsCollector *SnapshotterMetricsCollector
+var GlobalSnapshotterCacheUsageCollector *SnapshotterCacheUsageCollector
+
+func init() {
+	GlobalSnapshotterMetricsCollector = &SnapshotterMetricsCollector{}
+	GlobalSnapshotterCacheUsageCollector = &SnapshotterCacheUsageCollector{}
+}
+
+type SnapshotterMethod string
+
+const (
+	SnapshotterMethodUnknown SnapshotterMethod = "UNKNOWN"
+	SnapshotterMethodPrepare SnapshotterMethod = "PREPARE"
+	SnapshotterMethodMount   SnapshotterMethod = "MOUNT"
+	SnapshotterMethodRemove  SnapshotterMethod = "REMOVE"
+)
+
+func (s *SnapshotterMetricsCollector) Collect(snapshotID string, begin, end time.Time, method SnapshotterMethod) {
+	beginTime := begin.Format("2006-01-02 15:04:05.000")
+	endTime := end.Format("2006-01-02 15:04:05.000")
+	elapsed, _ := strconv.ParseFloat(fmt.Sprintf("%.6f", end.Sub(begin).Seconds()*1000), 64)
+
+	switch method {
+	case SnapshotterMethodPrepare:
+		data.PrepareTime.WithLabelValues(snapshotID, beginTime, endTime).Set(elapsed)
+	case SnapshotterMethodMount:
+		data.MountTime.WithLabelValues(snapshotID, beginTime, endTime).Set(elapsed)
+	case SnapshotterMethodRemove:
+		data.RemoveTime.WithLabelValues(snapshotID, beginTime, endTime).Set(elapsed)
+	case SnapshotterMethodUnknown:
+		fallthrough
+	default:
+		log.L.Warnf("Unknown method: %s", method)
+	}
+}
+
+func (s *SnapshotterCacheUsageCollector) Collect(path string) {
+	c, b := exec.Command("du", "-bs", path), new(strings.Builder)
```

Comment:
It relies on bash shell, please try `fs.DiskUsage`

Reply:
The function can be simplified like:

```
func (s *SnapshotterMetricsCollector) Collect(snapshotID string, begin, end time.Time, method SnapshotterMethod) {
	beginTime := begin.Format("2006-01-02 15:04:05.000")
	endTime := end.Format("2006-01-02 15:04:05.000")
	elapsed, _ := strconv.ParseFloat(fmt.Sprintf("%.6f", end.Sub(begin).Seconds()*1000), 64)

	var collector *prometheus.GaugeVec

	switch method {
	case SnapshotterMethodPrepare:
		collector = data.PrepareTime
	case SnapshotterMethodMount:
		collector = data.MountTime
	case SnapshotterMethodRemove:
		collector = data.RemoveTime
	case SnapshotterMethodUnknown:
		fallthrough
	default:
		log.L.Warnf("Unknown method: %s", method)
	}

	if collector != nil {
		collector.WithLabelValues(snapshotID, beginTime, endTime).Set(elapsed)
	}
}
```

Reply:
Got it. Thanks.